### PR TITLE
Stable Diffusion Dreambooth training

### PR DESCRIPTION
**Pull request by pcuenca**

### CODE REVIEW

**Comment by pcuenca** [comment]
We might want to put the images in a dataset.

File: stable_diffusion_dreambooth_training.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+* Dreambooth tends to overfit quickly. To get good quality images we have to find a 'sweet spot' between the number of training steps and the learning rate. We recommend to use a low learning rate and progressively increase the number of steps until results are satisfactory.
+* Dreambooth needs more training steps for faces. In our experiments, 800-1200 steps worked well when using a batch size of 2 and LR of 1e-6.
+* Prior preservation is important to avoid overfitting when training on faces. For other subjects it doesn't seem to make a huge difference.
+* If you see that the generated images are noisy, or the quality is degraded, it likely means overfitting. First, try the steps above to avoid it. If the generated images are still noisy, use the DDIM scheduler or run more inference steps (~100 worked well in our experiments).
+* EMA training doesn't seem to make a difference.
+* Training the text encoder in addition to the UNet has a big impact on quality. Our best results were obtained using a combination of text encoder fine-tuning, low LR, and a suitable number of steps. However, fine-tuning the text encoder requires a GPU with at least 24 GB of RAM.
```
Comment (patrickvonplaten) [inline stable_diffusion_dreambooth_training.md:54]: That's very interesting - maybe one more sentence here that even if:
- replacing ADAM with bits and bites 8-bit ADAM
- using gradient checkpointing 
- fp16
we won't be able to do 16GB would be helpful here to know IMO

File: stable_diffusion_dreambooth_training.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+## Learning Rate Impact
+
+Dreambooth overfits very quickly. To get good results it's important to tune the learning rate and the number of training steps in a way that makes sense for your dataset. In our experiments (detailed below), we fine-tuned on four different datasets with high and low learning rates. In all cases we were able to get better results with a low learning rate.
+
+## Experiments Settings
+
+All experiments were conducted using the [`train_dreambooth.py`](https://github.com/huggingface/diffusers/tree/main/examples/dreambooth) script with the `AdamW` optimizer on 2x 40GB A100s. We used the same seed and kept all hyperparameters equal across runs, except LR, number of training steps and the use of prior preservation.
```
Comment (patrickvonplaten) [inline stable_diffusion_dreambooth_training.md:62]: AdamW is very GPU memory intensive => wondering if we could bring 24GB down to 16GB somehow ....

**Review by patrickvonplaten — APPROVED** [review]
Nice and short - great write-up!

File: stable_diffusion_dreambooth_training.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+_Note: a previous version of this post was published [as a W&B report](https://wandb.ai/psuraj/dreambooth/reports/Dreambooth-Training-Analysis--VmlldzoyNzk0NDc3)_.
+
+## TL;DR: Recommended Settings
+
+* Dreambooth tends to overfit quickly. To get good quality images we have to find a 'sweet spot' between the number of training steps and the learning rate. We recommend to use a low learning rate and progressively increase the number of steps until results are satisfactory.
+* Dreambooth needs more training steps for faces. In our experiments, 800-1200 steps worked well when using a batch size of 2 and LR of 1e-6.
+* Prior preservation is important to avoid overfitting when training on faces. For other subjects it doesn't seem to make a huge difference.
+* If you see that the generated images are noisy, or the quality is degraded, it likely means overfitting. First, try the steps above to avoid it. If the generated images are still noisy, use the DDIM scheduler or run more inference steps (~100 worked well in our experiments).
+* EMA training doesn't seem to make a difference.
```
[the remark this answers is not in the document]
Comment (pcuenca) [reply stable_diffusion_dreambooth_training.md:53]: I think what we want to say is that fine-tuning with EMA enabled gives about the same results as without it. Is that correct @patil-suraj? Also, I'm not sure if we are starting with an EMA checkpoint, are we?

File: stable_diffusion_dreambooth_training.md (hunk 2 above)
Comment (pcuenca) [reply stable_diffusion_dreambooth_training.md:62]: That's a very good point. What do you think @patil-suraj?

File: _blog.yml [hunk 4]
```diff
@@ -1366,3 +1366,14 @@
     - ethics
     - research
     - nlp
+
+- local: stable_diffusion_dreambooth_training
```
Comment (osanseviero) [inline _blog.yml:1370]: A shorter name usually is nicer as this will be the url. E.g. `dreambooth`

File: _blog.yml [hunk 5]
```diff
@@ -1366,3 +1366,14 @@
...
+  title: "Training Stable Diffusion with Dreambooth using 🧨 Diffusers"
+  author: valhalla
+  thumbnail: /blog/assets/sd_dreambooth_training/thumbnail.png
+  date: Oct 30, 2022
```
Comment (osanseviero) [inline _blog.yml:1374]: To update

File: _blog.yml [hunk 6]
```diff
@@ -1366,3 +1366,14 @@
...
+  tags:
+    - diffusers
+    - stable-diffusion
+    - dreambooth
+    - fine-tuning
```
Comment (osanseviero) [inline _blog.yml:1379]: Also `guide`

File: stable_diffusion_dreambooth_training.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+title: Training Stable Diffusion with Dreambooth using Diffusers
+thumbnail: /blog/assets/sd_dreambooth_training/thumbnail.jpg
+---
+
+<h1>
+	Training Stable Diffusion with Dreambooth using 🧨 Diffusers
+</h1>
+
+<div class="blog-metadata">
+    <small>Published October 30th, 2022.</small>
```
Comment (osanseviero) [inline stable_diffusion_dreambooth_training.md:11]: To update

File: stable_diffusion_dreambooth_training.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        <div class="bfc">
+            <code>pcuenq</code>
+            <span class="fullname">Pedro Cuenca</span>
+        </div>
+    </a>
+	 <a href="/NineOfNein">
+        <img class="avatar avatar-user" src="https://pbs.twimg.com/profile_images/1262423730870988806/qA2FeHXF_400x400.jpg" width="100" title="Gravatar">
+        <div class="bfc">
+            <code>NineOfNein</code>
+            <span class="fullname">Valentine Kozin</span>
```
Comment (osanseviero) [inline stable_diffusion_dreambooth_training.md:36]: For guests we add a guest line as in https://github.com/huggingface/blog/blob/main/1b-sentence-embeddings.md?plain=1#L22

File: stable_diffusion_dreambooth_training.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+![Faces, prior preservation](https://api.wandb.ai/files/psuraj/images/projects/36980101/34f12faf.png)
+
+No prior preservation, 1200 steps, lr=`2e-6`.
+![Faces, prior preservation](https://api.wandb.ai/files/psuraj/images/projects/36980101/5e4e8b07.png)
+
+As you can see, results are better when prior preservation is used, but there are still noisy blotches. It's time for some additional tricks!
+
+## Effect of Schedulers
+
+In the previous examples we used the `PNDM` scheduler to sample images during the inference process. We observed that when the model overfits, `DDIM` usually works much better than `PNDM` and `LMSDiscrete`. In addition, quality can be improved by running inference for more steps: 100 seems to be a good choice. The additional steps help resolve some of the noise patches into image details.
```
Comment (osanseviero) [inline stable_diffusion_dreambooth_training.md:126]:
```suggestion
In the previous examples, we used the `PNDM` scheduler to sample images during the inference process. We observed that when the model overfits, `DDIM` usually works much better than `PNDM` and `LMSDiscrete`. In addition, quality can be improved by running inference for more steps: 100 seems to be a good choice. The additional steps help resolve some of the noise patches into image details.
```

File: stable_diffusion_dreambooth_training.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+## Epilogue: Textual Inversion + Dreambooth
+
+We also ran a final experiment where we combined [Textual Inversion](https://textual-inversion.github.io) with Dreambooth. Both techniques have a similar goal, but their approaches are different.
+
+In this experiment we first ran textual inversion for 2000 steps. From that model, we then ran Dreambooth for an additional 500 steps using a learning rate of `1e-6`. These are the results:
+
+![Textual Inversion + Dreambooth](https://api.wandb.ai/files/psuraj/images/projects/36980101/e25e766c.png)
+
+In our opinion, results are much better than doing plain Dreambooth, but not as good as when we fine-tune the whole text encoder. It seems to copy the style of the training images a bit more, so it could be overfitting to them. We didn't explore this combination further, but it could be a very interesting alternative to improve Dreambooth and still fit the process in a 16GB GPU. Feel free to explore and tell us about your results!
```
Comment (osanseviero) [inline stable_diffusion_dreambooth_training.md:168]:
```suggestion
We think the results are much better than doing plain Dreambooth but not as good as when we fine-tune the whole text encoder. It seems to copy the style of the training images a bit more, so it could be overfitting to them. We didn't explore this combination further, but it could be an interesting alternative to improve Dreambooth and still fit the process in a 16GB GPU. Feel free to explore and tell us about your results!
```

**Review by osanseviero — COMMENTED** [review]
Very cool and well written! I did a first pass mostly with nits and minor suggestions

**Comment by patrickvonplaten** [comment]
Actually one last thing, we could maybe add 1-2 sentences regarding our ethical thoughts on this, *e.g.* this should not be used to fine-tune a model that can generate harm etc...

**Comment by sayakpaul** [comment]
Crisp tutorial. 

Reminds of the canonical: https://youtu.be/myGAju4L7O8

Maybe the diffusers team could do a similar one.

**Comment by pcuenca** [comment]
Thanks a lot for the great suggestions and ideas! I did my best to incorporate them and I think it's ready to go, feel free to take another look if you want :)

**Review by patil-suraj — APPROVED** [review]
Very coo, thanks a lot for converting this into a blog !

**Comment by patrickvonplaten** [comment]
Cool looks good for me IMO

**Review by patrickvonplaten — APPROVED** [review]
